docs(core): add migration guide for static queries (#30644)
angular/angular: aio/content/guide/static-query-migration.md

# Static Query Migration Guide
​

> ****Attention to libraries: This migration is especially crucial for library authors to facilitate their users upgrading to version 9 when it becomes 

Reviewer: ```suggestion
****Important note for library authors: This migration is especially crucial for library authors to facilitate their users upgrading to version 9 when it becomes 
```

available (approx Oct 2019).****

In version 9, the default setting for `@ViewChild` and `@ContentChild` queries is changing in order to fix buggy and surprising behavior in queries 

> (see the section below for more details on that).

Reviewer: general comment: the md files typically follow the convention of not using new lines to break up lines. The rule is: "One line per one sentence" this is to make it easier to review git diffs for typo/minor tweaks to the language that could otherwise cause a refactoring of the full paragraph.

could you please refactor this to break the text so that each full sentence is on a single file? thanks

In preparation for this change, in version 8, we are migrating all applications and libraries to explicitly specify the resolution strategy for 
`@ViewChild` and `@ContentChild` queries.

Specifically, this migration adds an explicit "static" flag that dictates when that query's results should be assigned. Adding this flag will ensure 
your code works the same way when upgrading to version 9.   

Before:

```
// query results sometimes available in `ngOnInit`, sometimes in `ngAfterViewInit` (based on template)
@ViewChild('foo') foo: ElementRef; 
```

After:

```
// query results available in ngOnInit
@ViewChild('foo', {static: true}) foo: ElementRef; 

OR

// query results available in ngAfterViewInit
@ViewChild('foo', {static: false}) foo: ElementRef;
```

> Starting with version 9, the `static` flag will default to false. At this time, any `{static: false}` flags can be safely removed, and we will have a schematic 

Reviewer 2: I was confused whether `this time` refers to today, or to when version 9 is released, so maybe:

```suggestion
Starting with version 9, the `static` flag will default to false. At that time, any `{static: false}` flags can be safely removed, and we will have a schematic 
```

Also, removing the option is contradicting with it becoming required in #30639, is that meant as a temporary requirement to ensure that libraries specify it explicitly during the v8 lifecycle? If it is meant to become optional again in v9, I'd prefer I we mention that in the commit message of #30639.

Author: Ah, that's a good point re `this time`. Will update.

Re: removing the option: yes, the required status is completely temporary

Author: Update: @alxhub is updating his commit message for #30639

# Static Query Migration Guide
​
****Attention to libraries: This migration is especially crucial for library authors to facilitate their users upgrading to version 9 when it becomes 
available (approx Oct 2019).****

In version 9, the default setting for `@ViewChild` and `@ContentChild` queries is changing in order to fix buggy and surprising behavior in queries 
(see the section below for more details on that).

In preparation for this change, in version 8, we are migrating all applications and libraries to explicitly specify the resolution strategy for 
`@ViewChild` and `@ContentChild` queries.

> Specifically, this migration adds an explicit "static" flag that dictates when that query should be filled. Adding this flag will ensure your code works 

Reviewer 3: nit: "query should be filled" -> query result are assigned / set?

Author: Ah, yep, that's better, thanks

the same way when upgrading to version 9.   

Before:

```

> // query results sometimes available in `ngOnInit`, sometimes in `ngAfterViewInit` (based on template)

Reviewer 3: thinking out loud: would make sense to provide examples of 2 different templates?

Author: What would the second template look like?

that will update your code for you.

Note: this flag only applies to `@ViewChild` and `@ContentChild` queries specifically, as `@ViewChildren` and `@ContentChildren` queries do not have a concept of 
static and dynamic (they are always resolved as if they are "dynamic").

## FAQ

### What does this flag mean and why is it necessary?

The default behavior for queries has historically been undocumented and confusing, and has also commonly led to issues that are difficult to debug. In version 9, 
we would like to make query behavior more consistent and simple to understand. 

To explain why, first it's important to understand how queries have worked up until now.

Without the `static` flag, the compiler decided when each query would be resolved on a case-by-case basis. All `@ViewChild`/`@ContentChild` queries were 
categorized into one of two buckets at compile time: "static" or "dynamic". This classification determined when query results would become available to users.

- **Static queries** were queries where the result could be determined statically because the result didn't depend on runtime values like bindings. Results from 

> queries sorted as static were available before change detection ran for that view (accessible in `ngOnInit`).

Reviewer 3: nit: "queries sorted as static" - I must say that "sorted" never resonated with me in this context. Maybe queries "qualified" or "determined to be"?

Author: How about `classified`?

### Is there a case where I should use `{static: true}`?

This option was introduced to support creating embedded views on the fly. If you need to get ahold of a `TemplateRef` in a query to create a view dynamically, you won't be able 
to do so in `ngAfterViewInit`. Change detection has already run on that view, so creating a new view with the template will cause an `ExpressionHasChangedAfterChecked` error to 
be thrown. In this case, you will want to set the `static` flag to `true` and create your view in `ngOnInit`.

### ​Why generate {static: false}? Isn't that the default?

The point of this migration is to transition apps that aren't yet on v9 to a query pattern that is compatible with v9. However, most applications use libraries, and it's likely 
that some of these libraries may not be upgraded to version 8 (and thus might not have the proper flags). If we change the default, the behavior of queries in the library's 
components will change and possibly break. This way, an application's dependencies will behave the same during the transition. 

In Angular v9 and later, it will be safe to remove any `{static: false}` flags and we will do this cleanup for you in a schematic.

> ### Does this change affect `@ViewChildren` or `@ContentChildren` queries?

Reviewer 3: thinking out loud: pull this one up?

Author: Sure, where do you think it should go? First Q?

Author: Added as first Q

# Static Query Migration Guide
​

> ****Important note for library authors: This migration is especially crucial for library authors to facilitate their users upgrading to version 9 when it becomes available (approx Oct 2019).****

Reviewer: ```suggestion
**Important note for library authors: This migration is especially crucial for library authors to facilitate their users upgrading to version 9 when it becomes available (approx Oct 2019).**
```
